commons/src/main/java/org/infinispan/commons/api/functional/Closeables.java

```diff
@@ -0,0 +1,26 @@
+package org.infinispan.commons.api.functional;
+
+import java.util.Spliterator;
+
+// TODO: Remove once Will's branch closeable spliterator is in.
+
+/**
+ * @since 8.0
+ */
+public class Closeables {
+
+   private Closeables() {
+      // Cannot be instantiated, it's just a holder class
+   }
+
+   /**
+    * Auto closeable spliterator. Being able to close an spliterator is useful
+    * for situations where the iterator might consume considerable resources,
+    * e.g. in a distributed environment. Hence, being able to close it allows
+    * for resources to be reclaimed if the iterator won't be iterated any more.
+    */
+   public interface CloseableSpliterator<T> extends Spliterator<T>, AutoCloseable {
```

Comment:
I'm a bit confused where we're at with this discussion. In terms of Traversable, `spliterator` would return a auto-closeable spliterator exposing how different parts of the Traversable would be iterated in parallel, e.g. each segment could be iterated in parallel. A user could use this to iterate them in parallel in its own particular way.

In a normal JDK stream, the user has also the possibility to provide a spliterator on how data should be split, but not sure how useful it'd be in our case since we normally know how to best split data, e.g. by segments and operate over those in parallel.

With that in mind, maybe exposing `spliterator` does not make that much sense.

> The more I read, however, the more I'm thinking that we should offer a single lazy operation: stream(), which only uses resources (and releases them) when a "terminal operation" is invoked.

^ There's no single operation really here, there are several that expose Traversable back (ReadOnlyMap.entries...etc, ReadWriteMap.evalAll...etc). I think the suggestion could be good but the problem about it is that you then loose the option to close or cancel the Traversable before it's completed. In other words, if you're relying on terminal operations to complete to close all resources, you can't close it early.

If we implement your suggestion, to the make the API be consistent, we'd need to get rid of `iterator` and `spliterator`. The reason being that `iterator` and `spliterator` offer the possibility for you to stop iterating, but that would go against your suggestion to use terminal operations as a way to finish with resources.